Kashoo/synctos
# Issue 305: Separate modules for document constraint and document property validation

# Description

This decomposition results in easier to read and maintain code for document validation. It does not change the behaviour of generated sync functions.

# Testing

In addition to running the test cases with `npm test`, follow these steps for both Sync Gateway 1.x and 2.x to verify that it still produces sync functions that are valid:

1. Generate a sync function from some document definitions. For example:

```bash
./make-sync-function -j test/resources/time-doc-definitions.js build/test-sync-function.json
```

2. Copy and paste the contents of the sync function into a Sync Gateway configuration file.
3. Launch Sync Gateway with the configuration file.
4. Create a document that should fail validation. For example (note that, in this case, the "formatValidationProp" value is not a valid time and the "minAndMaxInclusiveValuesProp" value is greater than the maximum):

```
PUT /test/time-test HTTP/1.1
Host: localhost:4985
Content-Type: application/json

{ 
  "type": "timeDoc",
  "formatValidationProp": "24:00:00.001",
  "minAndMaxInclusiveValuesProp": "01:09:01.1"
}
```

5. Try again with a document that should pass validation. For example:

```
PUT /test/time-test HTTP/1.1
Host: localhost:4985
Content-Type: application/json

{ 
  "type": "timeDoc",
  "formatValidationProp": "24:00:00.001",
  "minAndMaxInclusiveValuesProp": "01:09"
}
```

# Related Issue

* GitHub issue link: #305
